### PR TITLE
backend testing integration with multiprocessing

### PR DESCRIPTION
**Pull request by RickSanchezStoic**
Adds multiprocessing for backend testing :

    test_function backend computation
    test_function ground truth computation 
    test_function gradient computation
    test_method  backend computation
    test_method ground truth computation
    test_method supported dtype and device calculation

### CODE REVIEW

File: ivy/data_classes/container/base.py [hunk 1]
```diff
@@ -4159,7 +4159,9 @@ def __getstate__(self):
         state_dict["_config_in"] = config_in
         config = copy.copy(state_dict["_config"])
         config["ivyh"] = (
-            config["ivyh"].current_backend_str() if config["ivyh"] is not None else None
+            config["ivyh"].current_backend_str()
+            if getattr(config, "ivyh", None) is not None
+            else None
```
Comment (CatB1t) [inline ivy/data_classes/container/base.py:4164]: Can we revert unrelated changes?

File: ivy_tests/test_ivy/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from .. import config
-
-if hasattr(config, "try_except"):
-    config.try_except()
```
Comment (CatB1t) [inline ivy_tests/test_ivy/__init__.py:4]: Can we revert unrelated changes?

File: ivy_tests/test_ivy/conftest.py [hunk 3]
```diff
@@ -2,24 +2,57 @@
...
+}  # multiversion
+mod_backend = {
+    "numpy": None,
+    "jax": None,
+    "tensorflow": None,
+    "torch": None,
+    "paddle": None,
+    "mxnet": None,
+}  # multiversion
+
```
Comment (CatB1t) [inline ivy_tests/test_ivy/conftest.py:26]: Why are those hard-coded values needed? can they be dynamic?

File: ivy_tests/test_ivy/conftest.py [hunk 4]
```diff
@@ -2,24 +2,57 @@
...
 UNSET_TEST_API_CONFIG = {"list": [], "flag": []}
 
 TEST_PARAMS_CONFIG = []
 UNSUPPORTED_FRAEMWORK_DEVICES = {"numpy": ["gpu", "tpu"]}
-
 if "ARRAY_API_TESTS_MODULE" not in os.environ:
     os.environ["ARRAY_API_TESTS_MODULE"] = "ivy.functional.backends.numpy"
 
 
+def default_framework_mapper(fw, set_too=False):
```
Comment (CatB1t) [inline ivy_tests/test_ivy/conftest.py:47]: Let's add a doc-string explaining what's going on/purpose of the function

File: ivy_tests/test_ivy/conftest.py [hunk 5]
```diff
@@ -2,24 +2,57 @@
...
+    # do a path search, get the latest one
+    versions = os.listdir(f"/opt/fw/{fw}")
```
Comment (CatB1t) [inline ivy_tests/test_ivy/conftest.py:49]: Ideally, we shouldn't use `/` but instead `os.path` so it can handle different path separators on different OS.

File: ivy_tests/test_ivy/conftest.py [hunk 6]
```diff
@@ -2,24 +2,57 @@
...
+    versions.sort()
+    if set_too:
+        sys.path.insert(1, f"/opt/fw/{fw}/{versions[-1]}")
```
Comment (CatB1t) [inline ivy_tests/test_ivy/conftest.py:52]: Ideally, we shouldn't use `/` but instead `os.path` so it can handle different path separators on different OS.

File: ivy_tests/test_ivy/conftest.py [hunk 7]
```diff
@@ -50,6 +83,81 @@ def pytest_configure(config):
...
+                input_queue.put(fw)
+                proc.start()
+
+                # we have the process running, the framework imported within,
+                # we now pack the queue and the process and store it in dict
+                # for future access
+                mod_backend[fw.split("/")[0]] = (proc, input_queue, output_queue)
+                # set the latest version for the rest of the test code and move on
+                default_framework_mapper(fw.split("/")[0], set_too=True)
+                found_backends.add(fw.split("/")[0])
```
Comment (CatB1t) [inline ivy_tests/test_ivy/conftest.py:110]: No need to do the splitting more than once, let's do it once and save the value for it, makes it clear to know what value is it. `truncated_fw=fw.split("/")[0]` for example.

File: ivy_tests/test_ivy/conftest.py [hunk 8]
```diff
@@ -50,6 +83,81 @@ def pytest_configure(config):
...
+        # no multiprocessing if multiversion
+        for fw in backend_strs:
+            if "/" in fw:
+                multiprocessing_flag = True
+                # multiversion, but without multiprocessing
+                sys.path.insert(1, f"/opt/fw/{fw}")
+
+    # frontend
+    frontend = config.getoption("--frontend")
+    if frontend:
```
Comment (CatB1t) [inline ivy_tests/test_ivy/conftest.py:145]:
```suggestion
    if frontend is not None:
```

File: ivy_tests/test_ivy/helpers/available_frameworks.py [hunk 9]
```diff
@@ -8,7 +9,10 @@ def _available_frameworks():
     for backend in ["numpy", "jax", "tensorflow", "torch", "paddle"]:
         if find_spec(backend) is not None:
             ret.append(backend)
-    return ret
+        elif os.path.exists(f"/opt/fw/{backend}"):
+            ret.append(backend)
+
+    return ["numpy", "jax", "tensorflow", "torch"]
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/available_frameworks.py:15]: again, ideally this shouldn't be hardcoded.

File: ivy_tests/test_ivy/helpers/available_frameworks.py [hunk 10]
```diff
@@ -8,7 +9,10 @@ def _available_frameworks():
     for backend in ["numpy", "jax", "tensorflow", "torch", "paddle"]:
         if find_spec(backend) is not None:
             ret.append(backend)
-    return ret
+        elif os.path.exists(f"/opt/fw/{backend}"):
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/available_frameworks.py:12]: `/opt/fw/` should be stored in a global variable as it seems being used widely.

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 11]
```diff
@@ -439,43 +352,223 @@ def test_function(
...
+        native_arrays,
+        container_flags,
+        none)
+    >>> fw = "numpy"
+    >>> fn_name = "add"
+    >>> x1 = np.array([1, 3, 4])
+    >>> x2 = np.array([-3, 15, 24])
+    >>> test_function(input_dtypes, test_flags, fw, fn_name, x1=x1, x2=x2)
+    """
+    fw = backend_to_test
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/function_testing.py:463]: No need to use the alias, let's use `backend_to_test` to make things straight forward.

File: ivy_tests/test_ivy/helpers/function_testing.py [hunk 12]
```diff
@@ -439,43 +352,223 @@ def test_function(
...
+    if mod_backend[fw]:
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/function_testing.py:464]: Doing that check every example is heavy, we should ideally bake the results before calling the pipeline.

File: ivy_tests/test_ivy/helpers/multiprocessing.py [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+    sys.path.insert(1, path)
+    framework = framework.split("/")[0]
+    framework = importlib.import_module(framework)
+    # if jax, do more stuff
+    if framework.__name__ == "jax":
+        framework.config.update("jax_enable_x64", True)
+
+    while True:
+        # subsequent arguments will be passed
+        data = input_queue.get()
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/multiprocessing.py:33]: Can we use a dict instead of a list?

**Review by CatB1t — CHANGES_REQUESTED** [review]
Hello @rishabgit, I think ideally we can re-think the paradigm of testing pipeline a bit, having a conditional statements everywhere in the testing pipeline isn't ideal, can we have a way of maybe baking the pipeline before starting so we know before-hand? for example

for no MP: 
`testing.build(no_mp_backend(), no_mp_gt_backend(), target="function_testing")`.

for mixed MP: 
`testing.build(no_mp_backend(),mp_gt_backend(), target="function_testing")`.

for full MP: 
`testing.build(mp_backend(),mp_gt_backend(), target="function_testing")`.

File: ivy/data_classes/container/base.py (hunk 1 above)
Comment (RickSanchezStoic) [reply ivy/data_classes/container/base.py:4164]: This is because of some corner case where `config["ivyh"]` is undefined as the key `ivyh` isn't present there, might be a mp thing, but not sure, the replaced code handles this corner case too.

File: ivy_tests/test_ivy/__init__.py (hunk 2 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/__init__.py:4]: This is because of some corner case where `config["ivyh"]` is undefined as the key `ivyh` isn't present there, might be a mp thing, but not sure, the replaced code handles this corner case too.

File: ivy_tests/test_ivy/conftest.py (hunk 3 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/conftest.py:26]: These have an important role, these dictionaries are used to pass on the `multiprocess` objects across the testing code.

File: ivy_tests/test_ivy/conftest.py (hunk 5 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/conftest.py:49]: We will be going to do the testing strictly using our docker image which will be a linux environment , so I guess we don't need to think about non-linux os here.

File: ivy_tests/test_ivy/conftest.py (hunk 7 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/conftest.py:110]: Noted

File: ivy_tests/test_ivy/helpers/available_frameworks.py (hunk 10 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/helpers/available_frameworks.py:12]: Yeah we sure can do that.

File: ivy_tests/test_ivy/helpers/available_frameworks.py (hunk 9 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/helpers/available_frameworks.py:15]: This was to adjust both the multiversion as well as  usual testing, we can refactor this a bit to come up with better logic.

File: ivy_tests/test_ivy/helpers/multiprocessing.py (hunk 13 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/helpers/multiprocessing.py:33]: You mean the data stream ? It needs to be a list or a tuple with indexing but I guess we can use a dictionary too, will need more refactoring of the code.

File: ivy_tests/test_ivy/helpers/function_testing.py (hunk 12 above)
Comment (RickSanchezStoic) [reply ivy_tests/test_ivy/helpers/function_testing.py:464]: I don't quite understand this one. This single if statement decides whether the testing here will be using mp or the usual case, so not sure how it adds overhead.

**Comment by RickSanchezStoic** [comment]
> Hello @RickSanchezStoic , I think ideally we can re-think the paradigm of testing pipeline a bit, having a conditional statements everywhere in the testing pipeline isn't ideal, can we have a way of maybe baking the pipeline before starting so we know before-hand? for example
> 
> for no MP: `testing.build(no_mp_backend(), no_mp_gt_backend(), target="function_testing")`.
> 
> for mixed MP: `testing.build(no_mp_backend(),mp_gt_backend(), target="function_testing")`.
> 
> for full MP: `testing.build(mp_backend(),mp_gt_backend(), target="function_testing")`.

We surely can! However, that would require some change to the already built pipeline. This one just adapts to the pipeline that's already there, with minimal changes. Will need to know some of the test building processes that occurs at the time.

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.
